### PR TITLE
Update `tsd` (and pin TypeScript to current minor)

### LINKED ISSUE #833: `ReadonlyDeep` / `WritableDeep` not modifying arrays as of TS 5.3.3
See a repro of the issue for `ReadonlyDeep` in this [TS Playground](https://www.typescriptlang.org/play?ts=5.3.3#code/KYDwDg9gTgLgBDAnmYcAKUCWBbTNMBuqAvAFBxwA+cAdgK4A2D5VcdNAJsAGaY3AcW1AM4wsNAOZDadbACNgUaXIgQGwAIY1pwxPLXLMEvjADcpUqEiwEyVACE6mBjACSNYXGLosufEVYCCEwOVgARDRhUagAlYAkAUXBzS3BoeCQUOAAJDWEAWUZ8MHUAYQ0mAGUjGki6KGBhAB4AFThQKM5PAAoAOn6NKAlZYBoYYQB9AC44LUQAbQBdAEovAD42GgBrGggAdxoNsgo2jtGOTwBvPoGhkbHJmb5uRTgAQWWZ9h39mlM4G69QbDbCjcbTWY0BYrL7bXYHAC+LAoAH5Nj8Dkt2iBOhd3siKHA0dwKsJgASKDMxHRyYS4DMSQwySkrOlbFk4hoOBAaAxEGFgMAwK0jnBTjjzp5HM43B4WGiWiwZuLcZ5+HsAf0gUNhDM5ktVsQNt94do6QqCcrsarNbcQWDHpDoYbjXDfhSiXBLgjrZK4JzubzEAB5OQAK2AAGMYAKhSKPaixQn6Tk8oUXJgSsBylUanUGs0Whszl0EFAacnE4q6TWU91AcD7uCZmhBhpQVEoIW1i7-cAYPUaC07CLVgAyPtcnl80MR6Ox4VFj1Wkt4gPTxBNddB-IaYXPV4AaWAiGHKAANHAD1A4AA1Co0s-ANZrZNo7d83dgBdNY+nuyXveDCPnYr61pSYq+qWH6bjBlT9k015wK4UTYE+L6Vp6cH9j+KHAGhoGYTMmEqn6DRTkGcBYtQ5GBnyVFavwRBQEsiyYYmtEblRbHgTWK4SqWnGUfMSEAKqXlqSG3jxvE1miQn0fMME-qJawSf0ymCsKt5rDJsmEvxNoKYgDH9GJl5SXp+mEvJmh0SZ8xappcaqZeznabp7F8ZBq6eDBbxQFAGibkheEESgYHWbW752RuAVBbBsVBrhqHoZFUV0oZfoQOGUYwF5vExRRM65fOWnxhl4GwhifwWJkqAwV+P5-k+gEPsA6FeF6PoTjBW5JZ+e79cV-LlS1oFuQNo1xkBIERS+KT1ZO9nwTG5VhZ13jenAvVTcNK0Ie5TQbaBC11XYy0brOeU-td0ZPlBeI5XOMCipcLDGVR41ZHwcBbCeEDcHAd0wE+iwzEdINPvM33AIsr4IikFjnVkABy7aNGAGiRiQXosN03CqDM9DyIonxwKI4gSOYFByBoABeMwqGomg0Es5iIxYkY8qIcAcJEGhde9FC1KCwhYzjMzbXk6KmrMnjo2LEvAOepCc6Q3MePAxkRDAGgQ1NP71YDfMC6K-N6+YOsC70ouY9jwC9HT9PmAA9K7lXgQAeiipDu8hNBipUcAAKy9AATBHMzdGAUAQCgsCIKszszJ9zPqFoSx+x77hB6HvQAMyF9Hsfx4oSDJwzTOqBnbNsUAA) (change the version to 5.2.2 to see the older correct behavior)

This is verified in the test here (though failing to fail as it should): https://github.com/sindresorhus/type-fest/blob/67f90a31a592352cc92002fe1e96e3a6d52ed276/test-d/readonly-deep.ts#L105

`WritableDeep` exhibits the same issue.

I suspect that the typing system is behaving differently in `tsd` than the raw `tsc` leading to this issue not being detected. Not sure why...

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/833">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/833/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by voxpelli**
As a preparation for #837 and as a companion to #838 I went ahead an updated `tsd` as well to the latest version, but turns out that the `tsd` tests are broken in latest version (likewise broken when rewriting to use eg. `expect-type` in latest typescript)

Fixes #833

### CODE REVIEW

**Comment by mrazauskas** [comment]
As far as I can understand, there are two failures here:

1.  As already mentioned in https://github.com/sindresorhus/type-fest/issues/833#issuecomment-2015222451, tests of `ReadonlyDeep` / `WritableDeep` are failing because of a change in TypeScript https://github.com/microsoft/TypeScript/pull/54448

2. The failure of `MergeDeep` test is a bug / feature of `tsd` (triggered by some another change in TypeScript).

   There are numerous issues in `tsd` repo pointing out that types are getting inferred different, if an expression is moved outside of an assertion. In this case, type of the `source` argument passed to `mergeDeep()` is inferred by `tsd` assertion as `(number | true)[]`:

   Screenshot 2024-03-22 at 21 24 46

   , but outside of assertion it is inferred as `(number | boolean)[]`:

   Screenshot 2024-03-22 at 21 21 03

   This difference if cause by the fact that `tsd` is inferring types already knowing what result is expected. More or less like here (and this can be interpreted both as a bug or a feature):

   Screenshot 2024-03-22 at 21 19 09

I guess you saw similar issues before and this is why `expect-type` got involved. The following assertion is passing:

```ts
expectTypeOf(mergeDeep(['life'], [42, true], {arrayMergeMode: 'spread'})).toEqualTypeOf<Array<number | string | boolean>>();
```

, because type is inferred correctly (in mine and `expect-type`s opinion):

Screenshot 2024-03-22 at 21 29 06

**Comment by voxpelli** [comment]
Seems like the issue was `HasMultipleCallSignatures`, it always returned `true`, but I made a fix for that

**Review by Emiyaaaaa — APPROVED** [review]

**Comment by voxpelli** [comment]
Needs one fix

**Comment by voxpelli** [comment]
Now it properly detects a multiple signature like this as well:

```ts
type Overloaded2 = {
	(foo: number | undefined): string;
	(foo: number): string;
};
```

File: tsconfig.json [hunk 1]
```diff
@@ -25,9 +25,5 @@
 		// "noUncheckedIndexedAccess": true, // TODO: Enable.
 		"noPropertyAccessFromIndexSignature": true,
 		"useDefineForClassFields": true,
-	},
-	"exclude": [
-		// Ignore `WriteableDeep` error temporarily, remove when #833 are fixed.
-		"test-d/writable-deep.ts",
```
Comment (sindresorhus) [inline tsconfig.json:31]: Would you be able to add a test to `test-d/writable-deep.ts` too? So it doesn't regress again in the future.

Comment (voxpelli) [reply tsconfig.json:31]: @sindresorhus Sure, there is a test already, but most tests in both of these are indirect ones which gives some cascading one, I'll add en explicit one 👍

Comment (voxpelli) [reply tsconfig.json:31]: @sindresorhus Done ✔️

**Comment by sindresorhus** [comment]
Great! Thanks 👍👍

**Pull request merged**